verl/utils/dataset/multiturn_sft_dataset.py

```diff
@@ -95,29 +119,67 @@ def __getitem__(self, item):
-
-            # Get tokens for messages up to previous point
-            prev_tokens = tokenizer.apply_chat_template(messages[:i], tools=tools, tokenize=True, return_tensors="pt", add_generation_prompt=False) if i > 0 else None
-
-            # Calculate start and end positions
-            start_pos = prev_tokens[0].shape[0] if prev_tokens is not None else 0
-            end_pos = prefix_tokens[0].shape[0]
-
-            # If this is an assistant message, set loss mask
-            if msg["role"] == "assistant":
+        i = 0
+        while i < len(messages):
+            cur_messages = messages[i]
+            if cur_messages["role"] == "assistant":
+                prev_applied_text = tokenizer.apply_chat_template(messages[:i], tokenize=False, add_generation_prompt=True, enable_thinking=enable_thinking)
+                cur_applied_text = tokenizer.apply_chat_template(messages[:i+1], tokenize=False, add_generation_prompt=False)
+                
+                # Get token positions for the assistant response
+                prev_tokens = tokenizer.encode(prev_applied_text, add_special_tokens=False)
+                cur_tokens = tokenizer.encode(cur_applied_text, add_special_tokens=False)
+                start_pos = len(prev_tokens)
+                end_pos = len(cur_tokens)
+                
+                # Set loss mask for assistant response
                 loss_mask[start_pos:end_pos] = 1
+                i += 1
+            elif cur_messages["role"] == "tool":
+                st = i
+                ed = i + 1
+                while ed < len(messages) and messages[ed]["role"] == "tool":
+                    ed += 1
+                prev_applied_text = tokenizer.apply_chat_template(messages[st:ed], tokenize=False, add_generation_prompt=False)
+                cur_applied_text = tokenizer.apply_chat_template(messages[st:ed+1], tokenize=False, add_generation_prompt=False)
+                
+                i = ed
+            elif cur_messages["role"] == "user":
+                prev_applied_text = tokenizer.apply_chat_template(messages[:i], tokenize=False, add_generation_prompt=False)
+                cur_applied_text = tokenizer.apply_chat_template(messages[:i+1], tokenize=False, add_generation_prompt=False)
+                i += 1
+            elif cur_messages["role"] == "system":
+                if i == 0:
+                    prev_applied_text = ""
+                else:
+                    raise ValueError("System message should be the first message")
+                    # prev_applied_text = tokenizer.apply_chat_template(messages[:i], tokenize=False, add_generation_prompt=False)
+                cur_applied_text = tokenizer.apply_chat_template(messages[:i+1], tokenize=False, add_generation_prompt=False)
+                i += 1
+            else:
+                raise ValueError(f"Unknown role: {cur_messages['role']}")
 
```

Comment:
I guess we can concatenate each token chunk and verify that it's the same as full_tokens. If it doesn't match, we can give user a warn as this is likely due to tokenizer issue.

Reply:
It also helps to make sure that the loss_mask is correct.

Reply:
Agree, a warning seems better than silence fail.